packages/cursorless-engine/src/tokenGraphemeSplitter/tokenGraphemeSplitter.ts

```diff
@@ -119,12 +119,26 @@ export class TokenGraphemeSplitter {
    * @param token The token to split
    * @returns A list of normalised graphemes in {@link token}
    */
-  getTokenGraphemes = (token: string): Grapheme[] =>
-    matchAll<Grapheme>(token, GRAPHEME_SPLIT_REGEX, (match) => ({
-      text: this.normalizeGrapheme(match[0]),
-      tokenStartOffset: match.index!,
-      tokenEndOffset: match.index! + match[0].length,
-    }));
+  getTokenGraphemes = (token: string): Grapheme[] => {
+    const graphemes = matchAll<Grapheme>(
+      token,
+      GRAPHEME_SPLIT_REGEX,
+      (match) => ({
+        text: this.normalizeGrapheme(match[0]),
+        tokenStartOffset: match.index!,
+        tokenEndOffset: match.index! + match[0].length,
+        isFirstLetterGrapheme: false,
+      }),
+    );
+    // iterate through the graphemes, marking the first letter
+    for (const grapheme of graphemes) {
+      if (grapheme.text.match(/[a-z]/)) {
+        grapheme.isFirstLetterGrapheme = true;
+        break;
+      }
+    }
```

Comment:
This doesn't feel like the right place for this logic; seems more specific to hat allocator.  I'd be inclined to inject it right after [we call `getTokenGraphemes` in `getTokenRemainingHatCandidates`](https://github.com/cursorless-dev/cursorless/blob/b0666d803adef38971019bd4d171956ca87dea8e/packages/cursorless-engine/src/util/allocateHats/allocateHats.ts#L141C1-L141C1) and add it to `HatCandidate`.

I'd also be tempted to handle camel case / snake case, but yeah prob not necessary in this draft? Shouldn't be too hard, tho; just for each char

1. if it's lowercase letter:
  a. return `false` if prev char exists and is letter
  b. else return `true`
2. else if it's uppercase letter:
  a. return `true` if previous char doesn't exist or is not uppercase letter
  b. else return `false`
3. else return `false`

I'd also advise using Unicode character classes, eg `\p{Lu}`, `\p{L}`, `\P{L}` etc

Also, it looks like your code should handle the case of the same grapheme appearing twice in the same token well, right?  Ie we'll end up with two candidates, and end up preferring the one that is not first in its token, right?

Reply:
Word splitting (for brains) is subtle. One case this would get wrong is languages that prefer all-caps initialisms, like Go and Swift. Consider [URLRequest](https://developer.apple.com/documentation/foundation/urlrequest): The second `R` begins a word, but would be marked as not-initial by your heuristic.

I was also worried about whether we needed to respect the settings from the user about how to tokenize.

But the perfect is the enemy of the good! I'll implement your suggestion and play with it.

Reply:
> I'd be inclined to inject it right after [we call getTokenGraphemes in getTokenRemainingHatCandidates](https://github.com/cursorless-dev/cursorless/blob/b0666d803adef38971019bd4d171956ca87dea8e/packages/cursorless-engine/src/util/allocateHats/allocateHats.ts#L141C1-L141C1) and add it to HatCandidate.

That seems reasonable. Will play with that. It'll spare me fixing 94 tests. :)

Reply:
Another hard word-splitting case: `thisIsATest`

Reply:
> I'd be inclined to inject it right after [we call getTokenGraphemes in getTokenRemainingHatCandidates](https://github.com/cursorless-dev/cursorless/blob/b0666d803adef38971019bd4d171956ca87dea8e/packages/cursorless-engine/src/util/allocateHats/allocateHats.ts#L141C1-L141C1)

At that point we've normalized the grapheme, so we no longer have case information. We could add an original text field to the graphemes. Or punt.

Reply:
Ahh right. Keep in mind tho that the normalisation won't always strip accents; the user can customise normalisation. You can use the offsets to get the original text

Your counterexamples for splitting are interesting, maybe better to just keep your heuristic?

But yeah I wonder if it's worth all the trouble 🤔

Maybe let's discuss at meet-up